files/default/neutron-ha-tool.py

```diff
@@ -753,24 +847,75 @@ def target_agent_list(agent_list, agent_type, exclude_agent_host):
     agent_info = agent_info[0]
     agent_mode = agent_info['configurations']['agent_mode']
 
-    return [agent['id'] for agent in agent_list
+    return [agent for agent in agent_list
             if agent['agent_type'] == agent_type and
             agent['alive'] and
             agent['host'] != exclude_agent_host and
             agent['configurations']['agent_mode'] == agent_mode]
 
 
-def agent_dead_id_list(agent_list, agent_type):
+def list_dead_agents(agent_list, agent_type):
     """
     Return a list of agents that are dead from an API list of agents
 
     :param agent_list: API response for list_agents()
 
     """
-    return [agent['id'] for agent in agent_list
+    return [agent for agent in agent_list
             if agent['agent_type'] == agent_type and agent['alive'] is False]
 
 
+class RandomAgentPicker(object):
+    def __init__(self, qclient, agents):
+        self.agents = agents
+
+    def pick(self):
+        return random.choice(self.agents)
+
+
+class LeastBusyAgentPicker(object):
+    def __init__(self, qclient, agents):
+        self.now = datetime.datetime.now
+        self.cache_created_at = None
+        self.qclient = qclient
+        self.agents_by_id = {agent['id']: agent for agent in agents}
+        self.router_count_per_agent_id = {}
+        self.refresh_router_count_per_agent_id()
+
+    def refresh_router_count_per_agent_id(self):
+        LOG.info("Refreshing router count per agent cache")
+        self.router_count_per_agent_id = dict()
+        for agent_id in self.agents_by_id:
+            self.router_count_per_agent_id[agent_id] = len(
+                list_routers_on_l3_agent(self.qclient, agent_id)
+            )
+        self.cache_created_at = self.now()
+
+    def cache_expired(self):
+        cache_life = self.now() - self.cache_created_at
+        return cache_life.seconds > ROUTER_CACHE_MAX_AGE_SECONDS
+
+    def pick(self):
+        if self.cache_expired():
+            self.refresh_router_count_per_agent_id()
+
+        agent_id_number_of_routers = sorted(
+            self.router_count_per_agent_id.items(),
+            key=lambda x: (x[1], x[0])
+        )
+        agent_id = agent_id_number_of_routers[0][0]
+        self.router_count_per_agent_id[agent_id] += 1
+        return self.agents_by_id[agent_id]
```

Comment:
Shouldn't we check that the agent is still alive? (it's actually not specific to this class, was also true for the random case) I can imagine the case where an agent dies, neutron-ha-tool is called to migrate things synchronously, but because it's synchronous, it takes some time and a second agent dies... Or is it purely theoretical?